internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs

```diff
@@ -20,10 +30,132 @@ impl RustArray {
     }
 }
 
+impl std::default::Default for RustArray {
+    fn default() -> Self {
+        RustArray {
+            array: std::ptr::null_mut(),
+            len: 0,
+            cap: 0,
+        }
+    }
+}
+
+impl From<Vec<u32>> for RustArray {
+    fn from(vec: Vec<u32>) -> Self {
+        RustArray::from_vec(vec)
+    }
+}
+
 #[no_mangle]
 pub extern "C" fn free_rust_array(array: RustArray) {
     let RustArray { array, len, cap } = array;
     unsafe {
         Vec::from_raw_parts(array, len, cap);
     }
 }
+
+#[repr(C)]
+pub enum Value {
+    None(()),
+    RustArray(RustArray),
+    U32(u32),
+    Ptr(*mut c_void),
+}
+
+macro_rules! impl_from_for_enum {
+    ($enum_name:ident, $($variant:ident => $type:ty),*) => {
+        $(
+            impl From<$type> for $enum_name {
+                fn from(value: $type) -> Self {
+                    $enum_name::$variant(value.into())
+                }
+            }
+        )*
+    };
+}
+
+impl_from_for_enum!(Value, None => (), RustArray => RustArray, RustArray => Vec<u32>, U32 => u32, Ptr => *mut c_void);
+
+#[repr(C)]
+pub struct RustResult {
+    pub success: bool,
+    pub value: Value,
+    pub error: *const c_char,
+}
+
+impl RustResult {
+    pub fn from_ptr(value: *mut c_void) -> Self {
+        RustResult {
+            success: true,
+            value: Value::Ptr(value),
+            error: std::ptr::null(),
+        }
+    }
+
+    pub fn from_error(error: String) -> Self {
+        RustResult {
+            success: false,
+            value: Value::None(()),
+            error: create_string(&error),
+        }
+    }
+}
+
+impl<T> From<Result<T>> for RustResult
+where
+    T: Into<Value>,
+{
+    fn from(value: error::Result<T>) -> Self {
+        match value {
+            Ok(v) => RustResult {
+                success: true,
+                value: v.into(),
+                error: null(),
+            },
+            Err(e) => RustResult {
+                success: false,
+                value: Value::None(()),
+                error: create_string(&e.to_string()),
+            },
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn free_rust_result(result: RustResult) {
+    match result.value {
+        Value::RustArray(array) => {
```

Comment:
In my view, the pointer type should also be freed.

Reply:
> In my view, the pointer type should also be freed.

Ptr type is catched by c++ class and will be freed when dtor is called for now. So we don't free ptr here.